File: firebase/functions/lib/events/live_meetings/breakouts/get_breakout_room_join_info.dart
import 'dart:async';

import 'package:firebase_functions_interop/firebase_functions_interop.dart';
import '../live_meeting_utils.dart';
import '../../../on_call_function.dart';
import '../../../utils/infra/firestore_utils.dart';
import 'package:data_models/cloud_functions/requests.dart';
import 'package:data_models/events/event.dart';
import 'package:data_models/events/live_meetings/live_meeting.dart';
import 'package:data_models/community/membership.dart';
import 'package:data_models/utils/utils.dart';

class GetBreakoutRoomJoinInfo
    extends OnCallMethod<GetBreakoutRoomJoinInfoRequest> {
  LiveMeetingUtils liveMeetingUtils;
  GetBreakoutRoomJoinInfo({LiveMeetingUtils? liveMeetingUtils})
      : liveMeetingUtils = liveMeetingUtils ?? LiveMeetingUtils(),
        super(
          'GetBreakoutRoomJoinInfo',
          (jsonMap) => GetBreakoutRoomJoinInfoRequest.fromJson(jsonMap),
        );

  @override
  Future<Map<String, dynamic>> action(
    GetBreakoutRoomJoinInfoRequest request,
    CallableContext context,
  ) async {
    final event = await firestoreUtils.getFirestoreObject(
      path: request.eventPath,
      constructor: (map) => Event.fromJson(map),
    );

    // Verify user is a participant
    final participant = await firestoreUtils.getFirestoreObject(
      path: '${request.eventPath}/event-participants/${context.authUid}',
      constructor: (map) => Participant.fromJson(map),
    );
    if (participant.status != ParticipantStatus.active) {
      throw HttpsError(HttpsError.failedPrecondition, 'unauthorized', null);
    }

    final liveMeetingPath =
        '${request.eventPath}/live-meetings/${request.eventId}';
    final liveMeeting = await firestoreUtils.getFirestoreObject(
      path: liveMeetingPath,
      constructor: (map) => LiveMeeting.fromJson(map),
    );

    final membershipDoc = await firestore
        .document(
          'memberships/${context.authUid}/community-membership/${event.communityId}',
        )
        .get();
    final membership = Membership.fromJson(
      firestoreUtils.fromFirestoreJson(membershipDoc.data.toMap()),
    );
    final isModOrCreator =
        event.creatorId == context.authUid || membership.isMod;

    // Verify user is a member of this breakout room
    final breakoutRoomDoc = await firestore
        .document(
          '$liveMeetingPath/breakout-room-sessions/${liveMeeting.currentBreakoutSession?.breakoutRoomSessionId}/breakout-rooms/${request.breakoutRoomId}',
        )
        .get();

    final breakoutRoom = BreakoutRoom.fromJson(
      firestoreUtils.fromFirestoreJson(breakoutRoomDoc.data.toMap()),
    );
    final isParticipantInBreakoutRoom =
        breakoutRoom.participantIds.contains(context.authUid);

    if (!isParticipantInBreakoutRoom && !isModOrCreator) {
      throw HttpsError(HttpsError.failedPrecondition, 'unauthorized', null);
    }

    final currentBreakoutSession = liveMeeting.currentBreakoutSession;
    if (currentBreakoutSession == null) {
      throw HttpsError(
        HttpsError.failedPrecondition,
        'no active breakout session',
        null,
      );
    }

    final breakoutRoomPath =
        '$liveMeetingPath/breakout-room-sessions/${currentBreakoutSession.breakoutRoomSessionId}/breakout-rooms/${request.breakoutRoomId}';

    final joinInfo = await liveMeetingUtils.getBreakoutRoomJoinInfo(
      communityId: event.communityId,
      eventId: request.eventId,
      breakoutSessionId: currentBreakoutSession.breakoutRoomSessionId ?? '',
      breakoutRoomPath: breakoutRoomPath,
      meetingId: breakoutRoom.roomId,
      userId: context.authUid!,
      record: breakoutRoom.record,
      existingRecordingSessionId: breakoutRoom.recordingSessionId,
      participantIds: breakoutRoom.participantIds,
    );

    // Map this user's Agora numeric UID to their Firebase user ID on the
    // recording session doc. The download-transcripts function uses this map
    // to replace raw Agora UIDs with display names in exported transcripts.
    //
    // We re-read the breakout room doc here because the local `breakoutRoom`
    // variable was fetched before `getBreakoutRoomJoinInfo` ran. If this is
    // the first user to join, `_startBreakoutRecording` will have created a
    // new recording session and written its ID to this doc in Firestore, but
    // our local snapshot still has the old (null) value.
    final updatedBreakoutSnap =
        await firestore.document(breakoutRoomPath).get();
    if (updatedBreakoutSnap.exists) {
      final updatedRoom = BreakoutRoom.fromJson(
        firestoreUtils.fromFirestoreJson(updatedBreakoutSnap.data.toMap()),
      );
      final sessionId = updatedRoom.recordingSessionId;
      if (sessionId != null) {
        await liveMeetingUtils.recordUidMapping(
          sessionId: sessionId,
          agoraUid: uidToInt(context.authUid!),
          userId: context.authUid!,
        );
      }
    }

    return joinInfo.toJson();
  }
}

File: firebase/functions/src/events/live_meetings/breakouts/get_breakout_room_join_info.ts
import * as functions from 'firebase-functions';
import { OnCallMethod } from '../../../on_call_function';
import { firestore, firestoreUtils } from '../../../utils/infra/firestore_utils';
import { Event, Membership, LiveMeeting, BreakoutRoom, BreakoutRoomSession, BreakoutRoomStatus, membershipIsMod, ParticipantStatus, Participant } from '../../../types';
import { LiveMeetingUtils } from '../live_meeting_utils';

interface GetBreakoutRoomJoinInfoRequest {
  eventPath: string;
  eventId: string;
  breakoutRoomId: string;
}

export class GetBreakoutRoomJoinInfo extends OnCallMethod<GetBreakoutRoomJoinInfoRequest> {
  private liveMeetingUtils: LiveMeetingUtils;

  constructor(liveMeetingUtils?: LiveMeetingUtils) {
    super('GetBreakoutRoomJoinInfo', (jsonMap) => jsonMap as GetBreakoutRoomJoinInfoRequest);
    this.liveMeetingUtils = liveMeetingUtils ?? new LiveMeetingUtils();
  }

  async action(request: GetBreakoutRoomJoinInfoRequest, context: functions.https.CallableContext): Promise<Record<string, unknown>> {
    const event: Event = await firestoreUtils.getFirestoreObject({
      path: request.eventPath,
      constructor: (map) => map as unknown as Event,
    });

    const participant: Participant = await firestoreUtils.getFirestoreObject({
      path: `${request.eventPath}/event-participants/${context.auth?.uid}`,
      constructor: (map) => map as unknown as Participant,
    });
    if ((participant as any).status !== ParticipantStatus.active) {
      throw new functions.https.HttpsError('failed-precondition', 'unauthorized');
    }

    const liveMeetingPath = `${request.eventPath}/live-meetings/${request.eventId}`;
    const liveMeeting: LiveMeeting = await firestoreUtils.getFirestoreObject({
      path: liveMeetingPath,
      constructor: (map) => map as unknown as LiveMeeting,
    });

    const membershipSnap = await firestore
      .doc(`memberships/${context.auth?.uid}/community-membership/${(event as any).communityId}`)
      .get();
    const membership = firestoreUtils.fromFirestoreJson(membershipSnap.data() ?? {}) as unknown as Membership;
    const isModOrCreator = (event as any).creatorId === context.auth?.uid || membershipIsMod((membership as any).status);

    const currentBreakoutSession = (liveMeeting as any).currentBreakoutSession as BreakoutRoomSession | undefined;
    const breakoutRoomDoc = await firestore
      .doc(
        `${liveMeetingPath}/breakout-room-sessions/${currentBreakoutSession?.breakoutRoomSessionId}/breakout-rooms/${request.breakoutRoomId}`
      )
      .get();

    const breakoutRoom = firestoreUtils.fromFirestoreJson(breakoutRoomDoc.data() ?? {}) as unknown as BreakoutRoom;
    const isParticipantInBreakoutRoom = (breakoutRoom as any).participantIds?.includes(context.auth?.uid);

    if (!isParticipantInBreakoutRoom && !isModOrCreator) {
      throw new functions.https.HttpsError('failed-precondition', 'unauthorized');
    }

    if (!currentBreakoutSession) {
      throw new functions.https.HttpsError('failed-precondition', 'no active breakout session');
    }

    const breakoutRoomPath = `${liveMeetingPath}/breakout-room-sessions/${currentBreakoutSession.breakoutRoomSessionId}/breakout-rooms/${request.breakoutRoomId}`;

    const joinInfo = await this.liveMeetingUtils.getBreakoutRoomJoinInfo({
      communityId: (event as any).communityId,
      eventId: request.eventId,
      breakoutSessionId: currentBreakoutSession.breakoutRoomSessionId ?? '',
      breakoutRoomPath,
      meetingId: (breakoutRoom as any).roomId,
      userId: context.auth!.uid,
      record: (breakoutRoom as any).record,
      existingRecordingSessionId: (breakoutRoom as any).recordingSessionId,
      participantIds: (breakoutRoom as any).participantIds ?? [],
    });

    return joinInfo as unknown as Record<string, unknown>;
  }
}
